extensions/notebook/src/book/bookTrustManager.ts

```diff
@@ -0,0 +1,89 @@
+		let trustedBookDirectory = this.getTrustedBookDirectory(notebookUri, workspace);
+
+		if (workspace && trustedBookDirectory) {
+			let trustedBook = this.getTrustedBook(normalizedWorkspacePath, trustedBookDirectory);
+
+			if (trustedBook) {
+				let fullBookBaseUri = path.join(normalizedWorkspacePath, trustedBookDirectory, 'content');
+				let requestingNotebookFormattedUri = notebookUri.substring(fullBookBaseUri.length).replace('.ipynb', '');
+				let notebookInTOC = trustedBook.tableOfContents.sections.find(jupyterSection => {
+					let normalizedJupyterSectionUrl = jupyterSection.url && path.normalize(jupyterSection.url);
+					return normalizedJupyterSectionUrl === requestingNotebookFormattedUri;
+				});
+				return !!notebookInTOC;
+			}
+		}
+
+		return false;
+	}
+
+	setBookAsTrusted(bookTreeItem: BookTreeItem): boolean {
+		// add this TOC to the configuration list
+		let workspacePathLength: number = path.resolve(vscode.workspace.rootPath).length;
+		let relativeBookPath: string = path.normalize(path.resolve(bookTreeItem.book.root).substring(workspacePathLength + 1));
+		let config: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration(BookTrustManager.notebookConfiguration);
+		let existingNotebooks: string[] = config[BookTrustManager.notebookTrustedBooksConfiguration];
+
+		// if no match found in the configuration, then add it
+		if (!existingNotebooks.find(notebookPath => path.normalize(notebookPath) === relativeBookPath)) {
+			existingNotebooks.push(relativeBookPath);
+
+			// update the configuration
+			config.update(BookTrustManager.notebookTrustedBooksConfiguration, existingNotebooks);
+			return true;
+		}
+		return false;
+	}
+
+	getTrustedBookDirectory(notebookUri: string, workspace: vscode.WorkspaceFolder): string {
+		let config: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration(BookTrustManager.notebookConfiguration);
+		let trustedBookDirectories: string[] = config.get(BookTrustManager.notebookTrustedBooksConfiguration);
+		let notebookUriWithoutBase: string = notebookUri.substring(path.normalize(workspace.uri.fsPath).length + 1);
+		return trustedBookDirectories.find(dir => notebookUriWithoutBase.startsWith(dir));
+	}
+
+	getTrustedBook(workspaceUri: string, baseBookUri: string): BookTreeItem {
+		let trustedBook = this.books
+			.map(book => book.bookItems) // select all the books
+			.reduce((accumulator, currentBookItemList) => accumulator.concat(currentBookItemList)) // flatten them to a single list
+			.find(bookTreeItem => {
+				let normalizedRootPath = path.normalize(bookTreeItem.root);
+				let fqnBookRootPath = path.join(workspaceUri, baseBookUri);
+				return normalizedRootPath.startsWith(fqnBookRootPath);
+			});
+		return trustedBook;
+	}
+
+	getNotebookWorkspaceFolder(notebookUri: string): vscode.WorkspaceFolder {
+		let workspace = vscode.workspace;
+		let workspaceFolder = workspace.workspaceFolders.find(wsFolder => {
```

Comment:
I don't believe books are always going to be opened as a workspace folder. Users can open books directly which just adds it to the tree view but doesn't actually open it in the workspace.